feat: Add PopoverNext component built on Floating UI (#7573)
palantir/blueprint: packages/core/src/components/popover-next/popover-next.md

its **target** (via the `renderTarget` prop or a single child element).

The **target** is rendered at the location of the PopoverNext component in the React component tree. It acts
as the trigger for the popover; user interaction will show the popover based on the `interactionKind` prop.
In Floating UI terms, this is the "reference" element. There are two ways to render a PopoverNext target, resulting
in different DOM layout depending on your application's needs:

-   The simplest way to specify a target is via `children`. Provide a single React child to
    `<PopoverNext>` and the component will render that child wrapped in a `@ns-popover-target` HTML element.
    This wrapper is configured with event handling logic necessary for the PopoverNext to function. Its tag name
    (e.g. `div`, `span`) and props can be customized with the `targetTagName` and `targetProps` props, respectively.

-   A more advanced API is available through the `renderTarget` prop. Here, PopoverNext provides you with all the
    information necessary to render a functional popover with a [render prop](https://reactjs.org/docs/render-props.html).

>     You are responsible for then propogating that information with an

Reviewer: ```suggestion
    You are responsible for then propagating that information with an
```

Author: fce1b5d261e5019223d8526d6db40ba9a1d9b03b

    [object spread](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals)

>     to the `React.JSX.Element` returned from `renderTarget`.

Reviewer: This sentence is quite confusing. You're responsible for spreading information into a JSX.Element? You should spread it into the props, not the resulting React object. This also sounds like you _call_ renderTarget, and then you spread the data to the returned object, but you actually return the object after spreading the props.

Author: Hm, yeah, another one that I copied over from the original Popover docs:
https://github.com/palantir/blueprint/blob/d8783f00b3ea2a637b62d6afcd9890d41a8614e3/packages/core/src/components/popover/popover.md?plain=1#L64-L66

I'll try to come up with some better wording here. fce1b5d261e5019223d8526d6db40ba9a1d9b03b

    -   If the rendered element is _not_ a native HTML element, you must ensure that it supports the

>         `className`, `ref`, and `tabIndex` props (i.e. renders them out to the DOM).

Reviewer: "into" maybe? Never heard "render out" before…
```suggestion
        `className`, `ref`, and `tabIndex` props (i.e. renders them into the DOM).
```

Author: fce1b5d261e5019223d8526d6db40ba9a1d9b03b

events, which interferes with the popover functioning. If you need to disable a button which
triggers a popover, you should use [`AnchorButton`](#core/components/button.anchor-button) instead.
See the [callout here](#core/components/button.props) for more details.

</div>

```tsx
import { Button, Classes, PopoverNext } from "@blueprintjs/core";

export const PopoverNextExample: React.FC = () => {
    // popover content gets no padding by default; add the "@ns-popover-content-sizing"
    // class to the popover to set nice padding between its border and content.
    return (
        <PopoverNext
            interactionKind="click"
            popoverClassName={Classes.POPOVER_CONTENT_SIZING}
            placement="bottom"
            content={
                <div>
                    <h5>Popover title</h5>
                    <p>...</p>
                    <Button className={Classes.POPOVER_DISMISS} text="Dismiss" />
                </div>
            }
            renderTarget={({ isOpen, ...targetProps }) => (
                <Button {...targetProps} intent="primary" text="PopoverNext target" />
            )}
        />
    );
};
```

@### Placement

> The `placement` prop controls the popover's position relative to the target. PopoverNext passes this prop directly to Floating UI; it uses the same semantics and supported values [as shown here in the docs](https://floating-ui.com/docs/computePosition#placement).

Reviewer: I think we should abstract this away. We can still refer to the Floating UI docs, but we should repeat the relevant parts (assuming the below is not complete) and remove "PopoverNext passes this prop directly to Floating UI"—that's implementation detail and may change in the future if they break API and we don't want to.

Author: I can get on board with that; first crack at breaking some of this out into our own docs explicitly: fce1b5d261e5019223d8526d6db40ba9a1d9b03b

<div class="@ns-callout @ns-intent-primary @ns-icon-info-sign">

You can also specify a specific initial placement (e.g. `"left"`, `"bottom-start"`) and still update the PopoverNext's position automatically by configuring the `flip` and `shift` middleware. [See below](#core/components/popover-next.middleware) for information about middleware.

</div>

@### Middleware

Middleware allow us to customize Floating UI's positioning behavior. **PopoverNext** configures several of Floating UI's built-in middleware to handle things such as flipping, preventing overflow from a boundary element, and positioning the arrow.

You may override the default middleware with the `middleware` prop, which is an object with key-value pairs representing the middleware name and its options object, respectively. See the [Floating UI middleware docs page](https://floating-ui.com/docs/middleware) for more info.

<div class="@ns-callout @ns-intent-warning @ns-icon-warning-sign @ns-callout-has-body-content">
    <h5 class="@ns-heading">Auto placement requires autoPlacement or flip middleware</h5>

> Be careful when disabling the "autoPlacement" or "flip" middleware, since automatic placement relies on them. If you _do_ decide to disable these middleware, be sure to also specify a placement which is specific (not automatic).

Reviewer: Do we need to support this API? It might be best to keep the simplest possible API for now until we know we need more, and even then maybe make it more semantic than just exposing all of Floating UI's API. That way, if we ever have to switch underlying libraries again, we might be able to get away without an API break next time.

Author: I get the concern about keeping the API simple and future-proof, but I'm thinking we should probably keep the middleware prop for now.

The main thing is that we're going to need this flexibility pretty much immediately. Looking at the current codebase, submenus and tooltips already rely on custom modifiers for things like `offset`, `flip`, and `preventOverflow`.

https://github.com/palantir/blueprint/blob/031a998e066f234067e46ca37d4e5ad192c51b42/packages/core/src/components/menu/menuItem.tsx#L308-L314

https://github.com/palantir/blueprint/blob/031a998e066f234067e46ca37d4e5ad192c51b42/packages/core/src/components/tooltip/tooltip.tsx#L126-L135

Without exposing middleware, we'd either have to recreate all these specific APIs (which feels like more work for less flexibility) or continually add on to functionality during the migration.

I also think the middleware concept is actually pretty well-designed already. Floating UI's API is well-documented, and it's tough to anticipate all the positioning edge cases we might need to handle. A more semantic API would probably just end up being a thin wrapper around middleware anyway, and then we'd be maintaining our own abstraction layer.

On the future-proofing front, if we do need to switch libraries again someday, the middleware concept seems pretty universal. Popper had modifiers, Floating UI has middleware, so there's probably a pattern here. We could potentially create a compatibility layer that maps between different systems if needed.

What's your take? Are there specific middleware use cases that concern you, or do you have a suggestion on a clean wrapper we could use here?

@## Style

@### Dark theme

**PopoverNext** automatically detects whether its trigger is nested inside a `.@ns-dark` container and applies the
same class to itself. You can also explicitly apply the dark theme to the React component by providing the prop
`popoverClassName="@ns-dark"`.

As a result, any component that you place inside a **PopoverNext** (such as a `Menu`) automatically inherits the dark theme
styles. Note that [`Tooltip`](#core/components/tooltip) uses **PopoverNext** internally, so it also benefits from this
behavior.

This behavior can be disabled (if the **PopoverNext** uses a **Portal**) via the `inheritDarkTheme` prop.

@### Sizing

Popovers by default have a `max-width` but no `max-height`. To constrain the height of a popover and make its content
scrollable, add a custom class to your popover content element and attach styles to that class:

```tsx
<PopoverNext content={<div className="custom-class">...</div>}>...</PopoverNext>
```

```scss
.custom-class {
    max-height: $pt-grid-size * 15;
    overflow-y: auto;
}
```

@reactExample PopoverNextSizingExample

@### Minimal style

You can create a minimal popover by setting `minimal={true}`.

> This removes the arrow from the popover and makes the transitions more subtle.

Reviewer: I believe we've talked about this one before, but do we want to split these two into separate props now that we're redoing the API anyway?

Author: Yeah, there are a few things to split out here:
- Presence of the arrow
- Offset of the popup element from the target
- Animation style

I've split `minimal` into two separate props, `arrow` and `animation`. The `arrow` prop controls the display of the arrow and applies the correct amount of offset to accommodate the arrow. The `animation` prop solely controls the style of the animation ("scale" vs. "minimal"). 86cd4c0b15ec6e6950d3d98a5b63e01e7672f7ac

---
tag: new
---

@# PopoverNext

PopoverNext displays floating content next to a target element.

The **PopoverNext** component is built on top of the [**Floating UI**](https://floating-ui.com) library.
Floating UI is a modern library that offers powerful, customizable, and performant positioning.

@## Import

```tsx
import { PopoverNext } from "@blueprintjs/core";
```

@reactExample PopoverNextExample

@## Usage

<div class="@ns-callout @ns-intent-warning @ns-icon-warning-sign @ns-callout-has-body-content">
    <h5 class="@ns-heading">

[OverlaysProvider](#core/context/overlays-provider) recommended

</h5>

This component renders an **Overlay2** which works best inside a React tree which includes an

> **OverlaysProvider**. Blueprint v5.x includes a backwards-compatibile shim which allows this context

Reviewer 2: **question**: Why do we need Overlay2? Is it for interactions with other overlays?

Reviewer 2: Interested to see porting this over in another implementation of this.

Author: The primary reasons we use [Overlay2](https://blueprintjs.com/docs/#core/components/overlay2) in the original Popover component are:
- Portal functionality
- Focus capturing
- Rendering transitions
- (optionally) Rendering backdrop components

The current set of functionality with Overlay2 was preserved in PopoverNext while porting the component, but it is worth noting that Floating UI can (theoretically) supplant Overlay2 with the features it exposes within the library itself:
- https://floating-ui.com/docs/FloatingPortal
- https://floating-ui.com/docs/FloatingFocusManager
- https://floating-ui.com/docs/useTransition
- https://floating-ui.com/docs/FloatingOverlay

This was out-of-scope for the original intent behind this PR to craft PopoverNext as a drop-in replacement, but noting here for future reference.
